Python/Product/ExternalProfilerDriver/VTuneToDWJSON.cs

```diff
@@ -209,6 +250,106 @@ class VTuneToDWJSON {
+                                                     ));
+
+            return mfdd;
+        }
+
+        public static Dictionary<string, List<SourceFileId>> SourceFilesByModule(Dictionary< string, Dictionary< string, FuncInfo > > modfun)
+        {
+
+            var modFile = modfun.Select( mfs => new { Module = mfs.Key, 
+                                                      FilesId = mfs.Value.Values.Select(fi => fi.SourceFile).Distinct().Where(f => File.Exists(f))
+                                                                         .Zip(Enumerable.Range(1, int.MaxValue), (x, y) => new SourceFileId(x, y) )
+                                                                         .ToList()
+                                                    })
+                                .Where(mfi => mfi.FilesId.Count > 0)
+                                .ToDictionary(mfi => mfi.Module, mfi => mfi.FilesId)
+                                ;
+
+            return modFile;
+        }
+
+        /// <summary>
+        /// Translates a two-level module -> func -> FuncInfo dictionary to the dwjson data model
+        /// </summary>
+        public static IEnumerable<ModuleSpec> ModFunToTrace(Dictionary< string, Dictionary< string, FuncInfo > > modfun)
+        {
+            var mfiles = SourceFilesByModule(modfun);
+            var mfilesDWJSON = mfiles.Select(kv => new { Module = kv.Key, 
+                                                         Files = kv.Value.Select(fi => new FileIDMapSpec {id = fi.Id, file = fi.SourceFile}).ToList()
+                                                       })
+                                     .ToDictionary(k => k.Module, v => v.Files);
+                                     
+            /*
+            This means that, to find the id that has been assigned to a filename in this structure we need something like:
+            var idx = mfilesDWJSON[<module>].FindIndex(fi => fi.file == <filename>);
+            var y = mfilesDWJSON[<module>][x];
+            */
+            
+            // Generate LineSpecs for each function in the module
+            List<ModuleSpec> modulesInTrace = new List<ModuleSpec>();
+            foreach (var mf in modfun.Select(mfd => new { Module = mfd.Key, Fnames = mfd.Value.Select(ffi => ffi.Value).ToList()})) {
+
+                ModuleSpec currentModule = new ModuleSpec {
+                    name = mf.Module
+                };
+                List<FunctionSpec> funcsInCurrentModule = new List<FunctionSpec>();
+                foreach (var ff in mf.Fnames) {
+                    //FunctionSpec currrent = new FunctionSpec (ff.FunctionName, ff.Base, ff.Size);
+                    FunctionSpec current = new FunctionSpec (ff.FunctionName, 0,0);
+
+                    // look up information on function
+                    if (mfilesDWJSON.ContainsKey(mf.Module)) {
+                        int idx = mfilesDWJSON[mf.Module].FindIndex(fi => fi.file == ff.SourceFile);
```

Comment:
Since you're just using this to get the file id anyway, try 
```
var idForFun = mfilesDWJSON[mf.Module].FirstOrDefault(fi => fi.file == ff.SourceFile)?.id;
if (idForFun.HasValue) { ... }
```